proposals/0000-template-patterns.rst

```diff
@@ -0,0 +1,376 @@
+
+Existing work on higher order unification has revealed that higher order unification of the template application form is a useful decidable subset of the general problem. This was first observed by Miller in `"Unification Under a Mixed Prefix" <https://repository.upenn.edu/cis_reports/454/>`_. Miller called these template applications 'patterns', but we decided against using that terminology because of the confusion with regular pattern matching in Haskell. Nowadays higher unification with Miller's pattern restriction is commonplace in dependently typed languages such as `Agda <http://www.cse.chalmers.se/~ulfn/papers/thesis.pdf>`_ and `Idris <https://www.type-driven.org.uk/edwinb/papers/impldtp.pdf>`_, which shows that higher order unification with this restriction is still very powerful.
+
+One notable extension of Miller's patterns is the extension by Duggan in his paper `"Unification with extended patterns" <https://doi.org/10.1016/S0304-3975(97)00141-2>`_. Duggan extends the simple template application form to allow projections of local binders as arguments to the template variables. Such an extension might also be applicable to the template application form in this proposal.
+
+Examples
+--------
+
+* 	One of the simplest examples is this rule:
+	::
+
+		{-# RULES "foo" forall f. foo (\x -> f x) = "RULE FIRED" #-}
+
+	It would match expressions like:
+	::
+
+		foo (\x -> x * 2 + x)
+
+* 	The higher order pattern may involve multiple locally bound variables, e.g.:
+	::
+
+		{-# RULES "foo" forall f. foo (\x y z -> f x y z) = "RULE FIRED" #-}
+
+	Which would match:
+	::
+
+		foo (\x y z -> x * y + z)
+
+	But not every variable has to occur in the match. It would also match this expression where ``y`` does not occur:
+	::
+
+		foo (\x y z -> x * 2 + z)
+
+* 	Locally bound variables may only occur once.
+	Consider the following rule:
+	::
+
+		{-# RULES "foo" forall f. foo (\x -> f x x) = "RULE FIRED" #-}
+
+	This would **not** match:
+	::
+
+		foo (\x -> x * 2 + x)
+
+	But it does contain the valid subrule ``f x``, so it would match:
+	::
+
+		foo (\x -> (bar x . baz) x)
+
+* 	Similarly if the template variable ``f`` is applied to non-variable arguments then it only matches a literal application.
+	Consider this rule:
+	::
+
+		{-# RULES "foo" forall f. foo (\x y -> f x 2 y) = "RULE FIRED" #-}
+
+	This would **not** match:
+	::
+
+		foo (\x y -> x * 2 + y)`
+
+	But again it does contain the higher order pattern ``f x``, so it would match:
+	::
+
+		foo (\x y -> (bar x . baz) 2 y)
+
+Effect and Interactions
+-----------------------
+
+The main effect of this proposal is that rewrite rules involving higher order patterns now match more expressions.
+But the additional matches are guaranteed to be beta equivalent, so this change does not cause existing rules to become semantically incorrect.
```

Comment:
What is it, beta or eta? I think eta. Or both? In what sense?

Reply:
I'm not sure what this means.  "Additional matches are beta equivalent".  What additional matches?  What does it mean for an additional match to be beta equivalent? @noughtmare ?

Reply:
I mean that anything that now matches under the new rules must be beta-equivalant to an expression that would have matched under the old rules.

Reply:
Currently we have a notion of syntactic matching, which means that the target must be syntactically equal to the template.

This proposal broadens/weakens that notion to semantic matching, so the target is allowed to be semantically beta-equivalent to the template in some cases.

So not all reason is lost.

Reply:
OK I have elaborated 3.1 to cover this.  See if you like it.